server/src/main/java/org/uiautomation/ios/wkrdp/WebKitNotificationListener.java

```diff
@@ -27,162 +35,278 @@
 import org.uiautomation.ios.wkrdp.message.WebkitApplication;
 import org.uiautomation.ios.wkrdp.message.WebkitPage;
 
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.List;
-import java.util.logging.Level;
-import java.util.logging.Logger;
-
 // TODO freynaud delete the interface ?
 public class WebKitNotificationListener implements MessageListener {
 
   private static final Logger log = Logger.getLogger(WebKitNotificationListener.class.getName());
+
   private final RemoteIOSWebDriver driver;
+
   private final ServerSideSession session;
+
   private final WebKitSynchronizer sync;
 
-  public WebKitNotificationListener(RemoteIOSWebDriver driver, WebKitSynchronizer syncronizer,
-                                    ServerSideSession session) {
+  private List<WebkitApplication> registeredApplications;
+
+  private Lock applicationRegistrationLock;
+
+  private Lock pagesProcessLock;
+
+  public WebKitNotificationListener(RemoteIOSWebDriver driver, WebKitSynchronizer syncronizer, ServerSideSession session) {
     this.driver = driver;
     this.session = session;
     this.sync = syncronizer;
+    this.registeredApplications = new ArrayList<>();
+    this.applicationRegistrationLock = new ReentrantLock();
+    this.pagesProcessLock = new ReentrantLock();
   }
 
   @Override
   public void onMessage(IOSMessage message) {
     if (message instanceof ReportSetupMessage) {
-      ReportSetupMessage m = (ReportSetupMessage) message;
-      driver.setDevice(m.getDevice());
-      sync.signalSimRegistered();
+      ReportSetupMessage reportSetupMessage = ReportSetupMessage.class.cast(message);
+      handle(reportSetupMessage);
+    } else if (message instanceof ReportConnectedApplicationsMessage) {
+      ReportConnectedApplicationsMessage reportConnectedApplicationsMessage = ReportConnectedApplicationsMessage.class
+          .cast(message);
+      handle(reportConnectedApplicationsMessage);
+    } else if (message instanceof ApplicationSentListingMessage) {
+      ApplicationSentListingMessage applicationSentListingMessage = ApplicationSentListingMessage.class.cast(message);
+      handle(applicationSentListingMessage);
+    } else if (message instanceof ApplicationDataMessage) {
+      // ApplicationDataMessage m = (ApplicationDataMessage)message;
+    } else if (message instanceof ApplicationConnectedMessage) {
+      ApplicationConnectedMessage applicationConnectedMessage = ApplicationConnectedMessage.class.cast(message);
+      handle(applicationConnectedMessage);
```

Comment:
I'm not a big fan of the `if` .. `else if` ... `else if` use here.  Can we re-factor this code a bit? 

Fo example;

1. `if` condition `{` 
        action; 
        return; 
     `}`  followed by more of the same

OR 

2. Extract method re-factor each condition and simply call each new method in this `onMessage(IOSMessage message)` method. Do the `instancecof` check in the new created methods.